---
title: "DExMA"
output: html_document
date: "2025-08-08"
---

# Setup
```{r setup, include=FALSE}
#Libraries
if (!requireNamespace("GenomicFeatures", quietly=TRUE))
  BiocManager::install("GenomicFeatures")
if (!requireNamespace("DexMA", quietly=TRUE))
  BiocManager::install("DExMA")


library(edgeR)
library(tibble)
library(tidyr)
library(dplyr)
library(pheatmap)
library(clusterProfiler)
library(enrichplot)
library(ggplot2)
library(org.Hs.eg.db)
library(forcats)
library(biomaRt)
library(GenomicFeatures)
library(DExMA)
library(RColorBrewer)
library(kableExtra)

knitr::opts_chunk$set(echo = FALSE, cache = FALSE)
```

# Directories
```{r directories}
# Path to the count tables for meta-analysis 
#dir_meta <- "/home/stefan/count-tables"

# Load count tables for meta-analysis
Khong_ars_G3BP1_counts <-     read.delim("/home/stefan/count-tables/Khong_counts.txt", sep = "\t", header = TRUE)
Matheny19_ars_EDC3_counts <- read.delim("/home/stefan/count-tables/Matheny19_counts.txt", sep = "\t", header = TRUE)
Iadevaia_vir_G3BP1_counts <- read.delim("/home/stefan/count-tables/Iadevaia_ars_G3BP1_counts.txt", sep = "\t", header = TRUE)
Somasekharan_counts <- read.delim("/home/stefan/count-tables/Somasekharan_counts.txt", sep = "\t", header = TRUE)
Curdy_ag_G3BP1_counts <- read.delim("/home/stefan/count-tables/Curdy_counts.txt", sep = "\t", header = TRUE)
Kudrin_ars_G3BP1_counts <- read.delim("/home/stefan/count-tables/Kudrin_NAT10ko_G3BP1_counts.txt", sep = "\t", header = TRUE)
Matheny21_ars_PABPC1_counts <- read.delim("/home/stefan/count-tables/Matheny21_ars_PABPC1_counts.txt", sep = "\t", header = TRUE)
Matheny21_sorb_G3BP1_counts <- read.delim("/home/stefan/count-tables/Matheny21_sorb_G3BP1_counts.txt", sep = "\t", header = TRUE)
Matheny21_sorb_PABPC1_counts <- read.delim("/home/stefan/count-tables/Matheny21_sorb_PABPC1_counts.txt", sep = "\t", header = TRUE)
Zhou_ars_G3BP1_counts <- read.delim("/home/stefan/count-tables/Zhou_ars_G3BP1_counts.txt", sep = "\t", header = TRUE)
Zhou_heat_G3BP1_counts <- read.delim("/home/stefan/count-tables/Zhou_heat_G3BP1_counts.txt", sep = "\t", header = TRUE)
Zhou_osm_G3BP1_counts <- read.delim("/home/stefan/count-tables/Zhou_osm_G3BP1_counts.txt", sep = "\t", header = TRUE)
Zhou_UV_puro_G3BP1_counts <- read.delim("/home/stefan/count-tables/Zhou_UV_puro_counts.txt", sep = "\t", header = TRUE)
Zhou_UV_G3BP1_counts <- read.delim("/home/stefan/count-tables/Zhou_UV_G3BP1_counts.txt", sep = "\t", header = TRUE)

Kudrin_G3BP1_NAT10ko_counts <- read.delim("/home/stefan/count-tables/Kudrin_NAT10ko_G3BP1_counts.txt", sep = "\t", header = TRUE)
Matheny21_sorb_G3BPko_counts <- read.delim("/home/stefan/count-tables/Matheny21_sorb_G3BPko_counts.txt", sep = "\t", header = TRUE)

```

# Meta Analysis Gene Expression
## DExMA Object creation 
```{r DExMA-ObjectMA}
###There is a how-to from Villatoro-GarcÃ­a: Differential Expression Meta-Analysis with DExMA package

# Create a list of the normalized count tables of the studies
listEX <- list(Khong_ars_G3BP1 = Khong_ars_G3BP1_counts,
               Matheny19_ars_EDC3 = Matheny19_ars_EDC3_counts,
               Iadevaia_vir_G3BP1 = Iadevaia_vir_G3BP1_counts,
               #Somasekharan = Somasekharan_counts,
               Curdy_ag_G3BP1= Curdy_ag_G3BP1_counts,
               Kudrin_ars_G3BP1 = Kudrin_ars_G3BP1_counts,
               Matheny21_ars_PABPC1 = Matheny21_ars_PABPC1_counts,
               Matheny21_sorb_G3BP1 = Matheny21_sorb_G3BP1_counts,
               Matheny21_sorb_PABPC1 = Matheny21_sorb_PABPC1_counts,
               Zhou_ars_G3BP1 = Zhou_ars_G3BP1_counts,
               Zhou_heat_G3BP1 = Zhou_heat_G3BP1_counts,
               Zhou_osm_G3BP1 = Zhou_osm_G3BP1_counts,
               Zhou_UV_puro_G3BP1 = Zhou_UV_puro_G3BP1_counts,
               Zhou_UV_G3BP1 = Zhou_UV_G3BP1_counts,
               Kudrin_G3BP1_NAT10ko = Kudrin_G3BP1_NAT10ko_counts,
               Matheny21_sorb_G3BPko = Matheny21_sorb_G3BPko_counts
               )

# Create pheno dataframes, where the two groups are specified
pheno1 <- data.frame(
  sampleID = colnames(Khong_ars_G3BP1_counts),
  condition = c("Input","Input","Input","SG", "SG", "SG")
)

pheno2 <- data.frame(
  sampleID = colnames(Matheny19_ars_EDC3_counts),
  condition = c("Input","Input","Input","SG", "SG", "SG")
)

pheno3 <- data.frame(
  sampleID = colnames(Iadevaia_vir_G3BP1_counts),
  condition = c("Input","Input","Input","SG", "SG", "SG")
)

pheno4 <- data.frame(
  sampleID = colnames(Somasekharan_counts),
  condition = c("Input","Input","Input","SG", "SG", "SG")
)

pheno5 <- data.frame(
  sampleID = colnames(Curdy_ag_G3BP1_counts),
  condition = c("Input", "Input", "Input", "Input", "Input", "SG", "SG", "SG", "SG", "SG")
)

pheno6 <- data.frame(
  sampleID = colnames(Kudrin_ars_G3BP1_counts),
  condition = c("Input", "Input", "Input", "Input", "SG", "SG", "SG", "SG")
)

pheno7 <- data.frame(
  sampleID = colnames(Matheny21_ars_PABPC1_counts),
  condition = c("Input", "Input", "Input", "SG", "SG", "SG")
)

pheno8 <- data.frame(
  sampleID = colnames(Matheny21_sorb_G3BP1_counts),
  condition = c("Input", "Input", "Input", "SG", "SG", "SG")
)

pheno9 <- data.frame(
  sampleID = colnames(Matheny21_sorb_PABPC1_counts),
  condition = c("Input", "Input", "SG", "SG")
)

pheno10 <- data.frame(
  sampleID = colnames(Zhou_ars_G3BP1_counts),
  condition = c("Input", "Input","Input", "SG", "SG", "SG")
)

pheno11 <- data.frame(
  sampleID = colnames(Zhou_heat_G3BP1_counts),
  condition = c("Input", "Input","Input", "SG", "SG", "SG")
)

pheno12 <- data.frame(
  sampleID = colnames(Zhou_osm_G3BP1_counts),
  condition = c("Input", "Input","Input", "SG", "SG", "SG")
)

pheno13 <- data.frame(
  sampleID = colnames(Zhou_UV_puro_G3BP1_counts),
  condition = c("Input", "Input","Input", "SG", "SG", "SG")
)

pheno14 <- data.frame(
  sampleID = colnames(Zhou_UV_G3BP1_counts),
  condition = c("Input", "Input","Input", "SG", "SG", "SG")
)

pheno15 <- data.frame(
  sampleID = colnames(Kudrin_G3BP1_NAT10ko_counts),
  condition = c("Input", "Input","Input", "Input", "SG", "SG", "SG", "SG")
)

pheno16 <- data.frame(
  sampleID = colnames(Matheny21_sorb_G3BPko_counts),
  condition = c("Input", "Input", "SG", "SG")
)




#Assign the rownames, so DExMA recognizes it
rownames(pheno1) <- pheno1$sampleID
rownames(pheno2) <- pheno2$sampleID
rownames(pheno3) <- pheno3$sampleID
rownames(pheno4) <- pheno4$sampleID
rownames(pheno5) <- pheno5$sampleID
rownames(pheno6) <- pheno6$sampleID
rownames(pheno7) <- pheno7$sampleID
rownames(pheno8) <- pheno8$sampleID
rownames(pheno9) <- pheno9$sampleID
rownames(pheno10) <- pheno10$sampleID
rownames(pheno11) <- pheno11$sampleID
rownames(pheno12) <- pheno12$sampleID
rownames(pheno13) <- pheno13$sampleID
rownames(pheno14) <- pheno14$sampleID
rownames(pheno15) <- pheno15$sampleID
rownames(pheno16) <- pheno16$sampleID

# Create list to assign pheno DFs to Studies
listPheno <- list(Khong_ars_G3BP1 =       pheno1,
                  Matheny19_ars_EDC3 =    pheno2,
                  Iadevaia_vir_G3BP1 =    pheno3,
                  #Somasekharan =          pheno4,
                  Curdy_ag_G3BP1 =        pheno5,
                  Kudrin_ars_G3BP1 =      pheno6,
                  Matheny21_ars_PABPC1 =  pheno7,
                  Matheny21_sorb_G3BP1  = pheno8,
                  Matheny21_sorb_PABPC1 = pheno9,
                  Zhou_ars_G3BP1 =        pheno10,
                  Zhou_heat_G3BP1 =       pheno11,
                  Zhou_osm_G3BP1 =        pheno12,
                  Zhou_UV_puro_G3BP1 =    pheno13,
                  Zhou_UV_G3BP1 =         pheno14,
                  Kudrin_G3BP1_NAT10ko =  pheno15,
                  Matheny21_sorb_G3BPko = pheno16
                  )

phenoGroups  <- c("condition", # 1
                  "condition", # 2
                  "condition", # 3
                  #"condition", # 4
                  "condition", # 5
                  "condition", # 6
                  "condition", # 7
                  "condition", # 8
                  "condition", # 9
                  "condition", # 10
                  "condition", # 11
                  "condition", # 12
                  "condition", # 13
                  "condition", # 14
                  "condition", # 15
                  "condition" # 16
                  )

phenoCases <- list(Khong_ars_G3BP1 =        "SG",
                   Matheny19_ars_EDC3 =     "SG",
                   Iadevaia_vir_G3BP1 =     "SG",
                   #Somasekharan =           "SG",
                   Curdy_ag_G3BP1 =         "SG",
                   Kudrin_ars_G3BP1 =       "SG",  
                   Matheny21_ars_PABPC1 =   "SG",
                   Matheny21_sorb_G3BP1  =  "SG",
                   Matheny21_sorb_PABPC1 =  "SG",
                   Zhou_ars_G3BP1 =         "SG",
                   Zhou_heat_G3BP1 =        "SG",
                   Zhou_osm_G3BP1 =         "SG",
                   Zhou_UV_puro_G3BP1 =     "SG",
                   Zhou_UV_G3BP1 =          "SG",
                   Kudrin_G3BP1_NAT10ko =   "SG",
                   Matheny21_sorb_G3BPko =  "SG"
                   )

phenoControls <- list(Khong_ars_G3BP1 =       "Input", 
                      Matheny19_ars_EDC3 =    "Input", 
                      Iadevaia_vir_G3BP1 =    "Input",
                      #Somasekharan =         "Input", 
                      Curdy_ag_G3BP1 =        "Input",
                      Kudrin_ars_G3BP1 =      "Input",  
                      Matheny21_ars_PABPC1 =  "Input",
                      Matheny21_sorb_G3BP1 =  "Input",
                      Matheny21_sorb_PABPC1 = "Input",
                      Zhou_ars_G3BP1 =        "Input",
                      Zhou_heat_G3BP1 =       "Input",
                      Zhou_osm_G3BP1 =        "Input",
                      Zhou_UV_puro_G3BP1 =    "Input",
                      Zhou_UV_G3BP1 =         "Input",
                      Kudrin_G3BP1_NAT10ko =  "Input",
                      Matheny21_sorb_G3BPko = "Input"
                      )

# Create ObjectMA
maObject <- createObjectMA(
  listEX     = listEX,
  listPheno  = listPheno,
  namePheno  = phenoGroups,
  expGroups  = phenoCases,
  refGroups  = phenoControls
)

#Check the ObjectMA. It is a list of nested lists. Each nested list is associated with one dataset and consists of two elements: The expression matrix [,[[1]] and a vector of 0 and 1s indicating controls (0) and cases (1) [,[[2]]
#str(maObject)

# Check the lists. Change the numbers to view the elements.
#maObject[[1]][[1]]
```


```{r Dataset_overview}
datasets_table <- read.delim("/home/stefan/supplementary_files/datasets_SER_SG.csv", sep = ";", header = TRUE)

datasets_table %>% kbl() %>% kable_styling()
```



## DExMA QC
```{r DExMA-QC}
# Log2 transformation
maObject <- dataLog(maObject)

#Test for heterogenity 
heterogeneityTest(maObject)


```

## DExMA Meta analysis
```{r DExMA-metaanalysis}
# Weights for Stouffer (DExMA calculates this automatically if the maObject is not null)
#w <- sapply(maObject, function(study) {
#  n1 <- sum(study$condition == 1)  
#  n0 <- sum(study$condition == 0)  
#  sqrt((n1 * n0) / (n1 + n0))
#})


# Perform Meta Analysis
resultsMA <- metaAnalysisDE(maObject, 
                            typeMethod = "Stouffer", 
                            missAllow= 0,       #maximum proportion of values allows in a sample
                            proportionData=0.8    # The proportion of datasets in which the gene should appear
                            )
```

## Heatmap
```{r DExMA-heatmap}

makeHeatmap(objectMA=maObject, 
            resMA=resultsMA, 
            scaling = "zscor", 
            regulation = "all", 
            numSig=500, 
            fdrSig = 0.05, 
            logFCSig = 0.5, 
            show_rownames = FALSE)



```

# Single dataset p values
```{r DExMA-pvals}

# Calculating indivdiual p-values and log2-FC
pvals <- pvalueIndAnalysis(maObject)

#pvals$logFC
#pvals$p
```


# Annotation and m6a data
```{r anno-m6a-data-loading}
# Load r_anno which was built in the SE/PE_processing script.
r_anno <- read.csv("r_anno.csv")

# Generate a column with Gene Names in order to merge() with annotation data
resultsMA.tib <- resultsMA %>%
  rownames_to_column(var = "ensembl_gene_id")

# Merge with r_anno
resultsMA.anno <- merge(
  resultsMA.tib,
  r_anno,
  by = "ensembl_gene_id",
  all.x = TRUE
)

#Calculate the gene length and add it as a column
resultsMA.anno <- resultsMA.anno %>%
  dplyr::mutate(gene_length = (end_position - start_position + 1), .after = end_position)

# Load table with m6A count data which was downloaded from the RMBase website (Download -> modGene -> Download Genes)
m6A_table <- read.delim("/data/stefan/RMbase_v3/gene.num.human.hg38.catAll.result.col29.splitNum.txt", sep = "\t", header = FALSE)

# Filter for only ENSG names in V1 and cut off version number. Filter for only ENSG name and m6a count
m6a_table.f <- m6A_table %>%
  filter(grepl("^ENSG", V1)) %>%
  mutate(V1 = sub("\\..*$", "", V1)) %>%
  dplyr::select(
    ensembl_gene_id = V1,
    m6aNum  = V6,
    Biotype = V4,
    Region = V5
  )


# Merge with DEG results table by gene_name
resultsMA.anno <- merge(
  resultsMA.anno,
  m6a_table.f,
  by = "ensembl_gene_id",
  all.x = TRUE
)

# Calculate m6A counts normalized by transcript_length (in kb)
resultsMA.anno.f <- resultsMA.anno %>%
                    mutate (
                      m6anum.n = (m6aNum / (transcript_length/1000))) %>%
                    filter(gene_biotype == "protein_coding")
                    
```

# Filter significant DEGs
```{r filter significant DEGs}
## Filter out significatly enriched and depleted by FC treshold
# subset upregulated genes from MA results with log2FC cutoff and p-value
resultsMA.anno.up  <- resultsMA.anno.f %>%
  filter(AveFC >= 0.5 & Pval <= 0.05 )
# subset all genes that are depleted
resultsMA.anno.down  <- resultsMA.anno.f %>% 
  filter(AveFC <= -0.5 & Pval <= 0.05)


# pre-ggplot2: Add class column 
up_df <- resultsMA.anno.up   %>% mutate(class = "SG")
dn_df <- resultsMA.anno.down %>% mutate(class = "Input")
universe_df <- resultsMA.anno.f %>% mutate(class = "Input") # Total RNA in this case is defined as unfiltered Meta DEG analysis output


plot_df <- bind_rows(up_df, universe_df) %>%
  dplyr::filter(class %in% c("Input","SG")) %>%
  dplyr::mutate(class = factor(class, levels = c("Input","SG"))) %>%
  dplyr::filter(!is.na(transcript_length),
                is.finite(transcript_length),
                transcript_length > 0) %>%   
  droplevels()

```

# Statistics
## Gene Length
```{r stat-gene-length}
# Create length statistics
length_summary <- plot_df %>%
  group_by(class) %>%
  summarise(
    n      = dplyr::n(),
    median = median(transcript_length),
    mean   = mean(transcript_length),
    IQR    = IQR(transcript_length),
    .groups = "drop"
)

length_summary

# Test length difference significance 
wilcox.test(transcript_length ~ class, data = plot_df)
```

## m6a count
```{r stat-m6a-count}
m6a_summary <- plot_df %>%
  group_by(class) %>%
  summarise(
    n      = dplyr::n(),
    median = median(m6anum.n, na.rm = TRUE),
    mean   = mean(m6anum.n, na.rm = TRUE),
    IQR    = IQR(m6anum.n, na.rm = TRUE),
    .groups = "drop"
    
)

m6a_summary
wilcox.test(m6anum.n ~ class, data = plot_df)
```

## Exon count
```{r Exon-count}

exon_summary <- plot_df %>%
  group_by(class) %>%
  summarise(
    n      = dplyr::n(),
    median = median(exon_count, na.rm = TRUE),
    mean   = mean(exon_count, na.rm = TRUE),
    IQR    = IQR(exon_count, na.rm = TRUE),
    .groups = "drop"
    
)

exon_summary
wilcox.test(exon_count ~ class, data = plot_df)


```

## Gene Biotype
```{r biotypes}

biotype_dexma <- resultsMA.anno %>%
  filter(AveFC >= 0.5 & Pval <= 0.05) %>%
  count(gene_biotype) %>%
  mutate(gene_biotype = ifelse(gene_biotype %in% c("protein_coding", "lncRNA"),
                               gene_biotype, "rest")) %>%
  group_by(gene_biotype) %>%
  summarise(n = sum(n), .groups = "drop") %>%
  mutate(percentage = n / sum(n) * 100) %>%
  mutate(Study = "DExMA")


biotype_biancon <- data.frame(
  gene_biotype = c("protein_coding", "lncRNA", "rest"),
  n            = c(89.5, 5.3, 5.2),
  percentage   = c(89.5, 5.3, 5.2),
  Study        = "Biancon et al."
)



biotype_all <- bind_rows(biotype_dexma, biotype_biancon)

ggplot(biotype_all, aes(x = Study, y = percentage, fill = gene_biotype)) +
  geom_bar(stat = "identity", width = 0.6,) +
  geom_text(
    aes(label = sprintf("%.1f%%", percentage)),
    position = position_stack(vjust = 0.5),  # center inside each segment
    color = "black", size = 3.5
  ) +
  scale_fill_brewer(palette = "Paired")  +
  labs(x = NULL, y = "Percentage", fill = "Biotype") +
  theme_classic() +
  theme(
    plot.margin  = margin(5, 5, 5, 5),
    panel.border = element_rect(color = "black", fill = NA, linewidth = 0.5)
  ) +
  scale_y_continuous(expand = c(0, 0.5), limits = c(0, 110))


```


# Outlier removal 
```{r remove-outliers}
hist(resultsMA.anno.f$m6anum.n, breaks=50)
hist(resultsMA.anno.f$transcript_length, breaks=50)

# Calculate cutoff for m6a sites
Q1_m6a <- quantile(resultsMA.anno.f$m6anum.n, 0.25, na.rm = TRUE)
Q3_m6a <- quantile(resultsMA.anno.f$m6anum.n, 0.75, na.rm = TRUE)
IQR_m6a <- Q3_m6a - Q1_m6a
cutoff_m6a <- Q3_m6a + 1.5 * IQR_m6a

# Calcualte cutoff for trans length
Q1_trans <- quantile(resultsMA.anno.f$transcript_length, 0.25, na.rm = TRUE)
Q3_trans <- quantile(resultsMA.anno.f$transcript_length, 0.75, na.rm = TRUE)
IQR_trans <- Q3_trans - Q1_trans
cutoff_trans <- Q3_trans + 1.5 * IQR_trans

resultsMA.anno.filt <- resultsMA.anno.f %>%
                        filter(m6anum.n <= cutoff_m6a) #%>%
                        # filter(transcript_length <= cutoff_trans)

hist(resultsMA.anno.filt$m6anum.n, breaks=50)
hist(resultsMA.anno.filt$transcript_length, breaks=50)

#Number of outliers removed
nrow(resultsMA.anno.f) - nrow(resultsMA.anno.filt)

```

# Comparison of Enriched Genes with Biancon et al. 

```{r Biancon_comparison_enrichedgenes}
#Gene List of Biancon et al.,2025 Supplementary Table 3
biancon_genes <- read.csv("/home/stefan/supplementary_files/biancon_genes.csv", sep = ";", header = TRUE)


biancon_genes_high <- biancon_genes %>%
  #filter(consensus_class == "high") %>% 
  distinct(gene_symbol)

enriched_genesMA <- resultsMA.anno.up %>%
  dplyr::select("external_gene_name") %>%
  distinct(external_gene_name)

common_genes <- dplyr::inner_join(
  biancon_genes_high, enriched_genesMA,
  by = c("gene_symbol" = "external_gene_name")
) %>% dplyr::distinct(gene_symbol)

#Significatly enriched genes compared with Gene List from Biancon with consensus class = High
tibble(
  Biancon   = nrow(biancon_genes_high),
  Meta_analysis  = nrow(enriched_genesMA),
  Common    = nrow(common_genes)
)

```

# Plots
## FC, length and m6a plot
```{r FCvslength-and-m6a-plot}

ggplot(resultsMA.anno.filt, aes(x = transcript_length, y = AveFC, color = m6anum.n)) +
  geom_point(alpha = 0.5, size = 1.5) +
  geom_smooth(method = "lm", se = FALSE, color = "black", size = 0.5) +
  scale_color_viridis_c(option = "plasma") +
  scale_x_log10() +
  labs(
    x = "Transcript length",
    y = "Fold Enrichtment",
    color = "Number of m6a sites"
  ) +
  theme_classic() +
  theme(panel.grid.minor = element_blank(), axis.line = element_line()) +
  geom_hline(yintercept=0, color = "grey", size=0.2, linetype = 2)

```

## Length plot
```{r length-plot2}
# Make sure class is factor and > 0
plot_gl <- plot_df %>%
  filter(!is.na(transcript_length), transcript_length > 0) %>%
  mutate(class = factor(class)) 

# Plot
ggplot(plot_gl, aes(x = class, y = transcript_length, fill = class)) +
  geom_violin(width = 0.9, trim = TRUE, alpha = 0.5, color = NA) +
  #geom_jitter(color="#E7C9C6", size=0.4, alpha=0.5, width = 0.095) +
  geom_boxplot(width = 0.18, outlier.shape = NA, alpha = 0.85) +
  stat_summary(fun = median, geom = "point", size = 2.2, color = "black") +
  scale_fill_manual(values = c(Input = "#D3BA68", SG = "#65A479")) +
  scale_y_continuous(
    trans = "log10",
     labels = scales::label_number()
  ) +
  labs(x = NULL, y = "Transc. length [bp] (log)") +
  theme_classic() +
  theme(panel.grid.minor = element_blank(), legend.position = "none", axis.line = element_line()) +
  annotate("text", x = "SG", y = 100000, label = "****", size = 5)




 


```

## m6a plot
```{r m6a-plot2}


plot_gl <- plot_df %>%
  filter(!is.na(m6anum.n), m6anum.n > 0) %>%
  mutate(class = factor(class))  # set order if you like: levels=c("Input","SG")

ggplot(plot_gl, aes(x = class, y = m6anum.n, fill = class)) +
  geom_violin(width = 0.9, trim = TRUE, alpha = 0.5, color = NA) +
  geom_boxplot(width = 0.18, outlier.shape = NA, alpha = 0.85) +
  stat_summary(fun = median, geom = "point", size = 2.2, color = "black") +
  scale_fill_manual(values = c(Input = "#D3BA68", SG = "#65A479")) +
  scale_y_continuous(
    trans = "log10",
    labels = scales::label_number(),
    limits = c(1,1e3),
    expand = c(0, 0)
  ) +
  labs(x = NULL, y = "m6a sites [ /kbp transcript length] (log)") +
  theme_classic() +
  theme(panel.grid.minor = element_blank(), legend.position = "none", axis.line = element_line()) +
  annotate("text", x = "SG", y = 900, label = "n.s.", size = 3)


```

## Exon count plot
```{r exon-count-plot}
plot_gl <- plot_df %>%
  filter(!is.na(exon_count), exon_count > 0) %>%
  mutate(class = factor(class))  # set order if you like: levels=c("Input","SG")

ggplot(plot_gl, aes(x = class, y = exon_count, fill = class)) +
  geom_violin(width = 0.9, trim = TRUE, alpha = 0.5, color = NA) +
  geom_boxplot(width = 0.18, outlier.shape = NA, alpha = 0.85) +
  stat_summary(fun = median, geom = "point", size = 2.2, color = "black") +
  scale_fill_manual(values = c(Input = "#D3BA68", SG = "#65A479")) +
  coord_cartesian(ylim = c(0, 50)) + 
  labs(x = NULL, y = "Exon Count") +
  theme_classic() +
  theme(panel.grid.minor = element_blank(), legend.position = "none", axis.line = element_line()) +
  annotate("text", x = "SG", y = 50, label = "****", size = 5)
```


## Most enriched genes

```{r most-enriched-genes}
#  Order meta analysis results by Average Fold Change in descending order and isolate the first 10 (= 10 most enriched) ensembl_gene_ids
most_enriched <- resultsMA.anno %>% 
  arrange(desc(AveFC)) %>%
  dplyr::slice(1:10)

# Pull out the ensemble_gene_ids or "External gene name" as character vector  
#most_enriched_vec <- most_enriched$ensembl_gene_id
most_enriched_vec <- most_enriched$external_gene_name

# Pull out the average FC of the meta analysis
AveFC <- most_enriched$AveFC

# Make a dataframe to be able to filter out the most enriched genes using the character vector from above
pvals.FC_df <- as.data.frame(pvals$logFC)

#"Translate" to gene_symbols
rownames(pvals.FC_df) <- r_anno$external_gene_name[match(rownames(pvals.FC_df), r_anno$ensembl_gene_id)]
 
# Index by most_enriched and keeping the order of decending AveFC
pvals.FC_df_10 <- pvals.FC_df[most_enriched_vec, , drop = FALSE]

#Add the average FC to the 10 most enriched genes
pvals.FC_df_10 <- pvals.FC_df_10 %>%
  mutate(AveFC = AveFC)
```


```{r most-enriched-genes-plot, fig.height=6}
pvals_long <- pvals.FC_df_10 %>%
  rownames_to_column("gene") %>% 
  pivot_longer( cols = -c(gene, AveFC), names_to = "dataset", values_to = "logFC" ) %>%
  mutate(
    gene = factor(gene, levels = rev(rownames(pvals.FC_df_10))), # Use rev() here because coord_flip() will turn the list also upside down    
    dataset = factor(dataset, levels = unique(dataset)) 
  )

# split 
dataset_dots <- pvals_long %>%
  filter(dataset != "AveFC")

aveFC_dots <- pvals_long %>%
  distinct(gene, AveFC) %>%
  mutate(
    AveFC = as.numeric(AveFC),
    gene  = factor(gene, levels = levels(pvals_long$gene))
  )


# Plot 
ggplot(dataset_dots, aes(x = gene, y = logFC)) +
  geom_point(aes(fill = dataset, color = dataset),
              width = 0.2, height = 0, size = 2, alpha = 0.9,
              shape = 21, stroke = 0.4) +
   geom_point(data = aveFC_dots,
             aes(x = gene, y = AveFC),
             inherit.aes = FALSE,
             shape = 4, size = 4, stroke = 1.2, color = "red") +
  labs(x = "Gene", y = "logFC", fill = "Dataset", color = "Dataset") +
  theme_classic() +
  geom_hline(yintercept=0, color = "black", size=0.2, linetype = 2) +
  coord_flip() 

```

## GO Enritchment

```{r GO_enrichtment}

#Gene Ontology Enritchment = Biological Processses
go_bp <- enrichGO(
  gene          = resultsMA.anno.up$ensembl_gene_id,     
  OrgDb         = org.Hs.eg.db,
  keyType       = "ENSEMBL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,           
  qvalueCutoff  = 0.2             
)

#Gene Ontology Enritchment = Molecular Function 
go_mf <- enrichGO(
  gene          = resultsMA.anno.up$ensembl_gene_id,     
  OrgDb         = org.Hs.eg.db,
  keyType       = "ENSEMBL",
  ont           = "MF",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,           
  qvalueCutoff  = 0.2             
)

#Gene Ontology Enritchment = Cellular component
go_cc <- enrichGO(
  gene          = resultsMA.anno.up$ensembl_gene_id,     
  OrgDb         = org.Hs.eg.db,
  keyType       = "ENSEMBL",
  ont           = "CC",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,           
  qvalueCutoff  = 0.2             
)

dotplot(go_bp)
dotplot(go_mf)
dotplot(go_cc)

```

```{r volcano-plot}

pvals.p_df <- as.data.frame(pvals$p)
































